src/inmanta/references.py

```diff
@@ -534,6 +535,37 @@ def run(self, logger: "handler.LoggerABC") -> None:
         dict_path_expr.set_element(self.resource, value)
 
 
+@typing.runtime_checkable
+class MaybeReference(typing.Protocol):
+    """
+    DSL value that may represent a reference in the Python domain, while having a different value in the DSL domain.
+    """
+
+    def is_reference(self) -> Optional[Reference]:
```

Comment:
```suggestion
    def as_reference(self) -> Optional[Reference]:
```

from a method with is, I expect a boolean

Reply:
Yeah, I struggled with the naming for this one. I'm not sure about `as`, because it's really about what this object already represents (is). But I'll let it sink in a bit, it's probably the best option anyway.

Reply:
I added it as a protocol because `ast.type` doesn't know and doesn't need to know (import loops) about instances, only that some values may represent references even if they aren't an instance of `Reference`.

I'm not too attached too it, and I do see there are two sides to the story. But overall I like that it decouples us from `Instance` here.

Reply:
I would add that as a comment

Reply:
When reviewing, please keep in mind that this makes `inmanta.execute.util` quite a bit heavier, due to the `inmanta.references` import. But I still believe it appropriate. It then pulls in (recursively) `references`, `util`, `dict_path`. I think that's all conceptually allowed underneath `inmanta.execute.util`.

Reply:
Ah no, `references` also pulls in `inmanta.resources`, which is not appropriate under `inmanta.execute.util`, and causes all manner of import loops. I'm putting it back in `inmanta.references`, even if I find that it only half fits there.